test/src/folding/parsers/java_parser_test.dart

```diff
@@ -0,0 +1,390 @@
+
+void method() {} // Not the only thing in the line  8
+// Single                                           9
+void method2 // Not the only thing in the line      10
+// Single                                           11
+() {}
+}
+// Single                                             14
+// Single                                             15''',
+        expected: [
+          _FB(firstLine: 0, lastLine: 1, type: _T.singleLineComment),
+          _FB(firstLine: 2, lastLine: 13, type: _T.braces),
+          _FB(firstLine: 4, lastLine: 6, type: _T.singleLineComment),
+          _FB(firstLine: 14, lastLine: 15, type: _T.singleLineComment),
+        ],
+      ),
+
+      _Example(
+        'Imports at beginning',
+        code: '''
+package com.akvelon.temp;
+
+import java.util.Arrays;
+
+/* */
+/* */
+import java.util.Date;
+
+/*
+*/
+import java.lang.Math;
+
+public class MyClass {}''',
+        expected: [
+          _FB(firstLine: 0, lastLine: 6, type: _T.imports),
+          _FB(firstLine: 4, lastLine: 5, type: _T.singleLineComment),
+          _FB(firstLine: 8, lastLine: 9, type: _T.multilineComment),
+        ],
+      ),
+
+      _Example(
+        'Imports at mid and the end, no newline in the end.',
+        code: '''
+//
+import java.util.Arrays;
+import java.util.Date;
+public class MyClass {}
+import java.lang.Math;
+import java.lang.Exception;''',
+        expected: [
+          _FB(firstLine: 1, lastLine: 2, type: _T.imports),
+          _FB(firstLine: 4, lastLine: 5, type: _T.imports),
+        ],
+      ),
+
+      _Example(
+        'Comment after a closing brace.',
+        code: '''
+class MyClass {
+void method() {
+}// comment
+// comment
+}''',
+        expected: [
+          _FB(firstLine: 0, lastLine: 4, type: _T.braces),
+          _FB(firstLine: 1, lastLine: 2, type: _T.braces),
+        ],
+      ),
+
+      _Example(
+        'Multiline comment after a closing brace with following single line comment',
+        code: '''
+class MyClass {
+void method() {
+}/* comment */
+// comment
+}''',
+        expected: [
+          _FB(firstLine: 0, lastLine: 4, type: _T.braces),
+          _FB(firstLine: 1, lastLine: 2, type: _T.braces),
+        ],
+      ),
+
+      _Example(
+        'Service comment sequences do not form a foldable block.',
+        code: '''
+class MyClass {
+// [START section1]
+// [END section2]
+}''',
+        expected: [
+          _FB(firstLine: 0, lastLine: 3, type: _T.braces),
+        ],
+      ),
+      _Example(
+        'Pair symbols should be ignored inside a multiline comment',
+        code: '''
+/*[([       0
+])]*/       1
+''',
+        expected: [
+          _FB(firstLine: 0, lastLine: 1, type: _T.multilineComment),
+        ],
+      ),
+      _Example(
+        'Multiline comment after a brace',
+        code: '''
+class MyClass{            // 0
+
+} /*                         2
+* some weird comment         3
+*/                        // 4
```

Comment:
Why they are not joined in a union?

Reply:
I was also confused.
But that's how Highlight parser does this.
I am trying to make Fallback parser's logic as close to Highlight parser as possible